Content attachments in Nessie

snazy opened the pull request:
Allow binary values related to content-ids via database-adapter.

This is a prerequisite to have Icebert table and view metadata in Nessie.

The content attachment keys is a composite of:
* content-id
* object-type (some string, e.g. `snapshot` or `schema`)
* object-id (some identifier, e.g. a hash over the content value or some
  other identifier)

Attachment values:
* persisted with "format" _and_ "compression" attributes.
* field "object ID", which represents e.g. the snapshot ID
* "version" field for optimistic locking + CAS operations

The new functionality via `DatabaseAdapter` allows:
* unconditional bulk writes and reads, used to allow bulk writes + reads of
  for example the Iceberg snapshot + partition spec + schema for an Iceberg
  table.
* conditional writes (think: "insert" + "update if") to store per-content-id
  information about a table - for Iceberg this would be for example the
  "last sequence number" or "last column id".
* listing all attachment keys for a content id, which may return keys for
  non-existent attachments (compromise for non-transactional databases).

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/BinaryContent.java:
```diff
@@ -0,0 +1,58 @@
+  /** Potentially compressed binary data. */
+  ByteString getData();
+
+  /** Optional "hash" value, used for optimistic locking. */
```
Optional in what sense? Do we not control it completely on the server side?

snazy replied:
The issue here, actually: where I'm not yet sure about, is whether it can/should have a "proper hash" or whether some random number is sufficient. So the term "hash" is probably wrong.

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/BinaryKey.java:
```diff
@@ -0,0 +1,44 @@
+  /** Content ID to which the binary object identified by this key belongs to. */
+  ContentId getContentId();
+
+  /** The object type, for example {@code snapshot} or {@code schema}. */
+  String getObjectType();
+
+  /** The object ID, for example the snapshot-ID or schema-ID. */
+  String getObjectId();
```
Do we expect "object ID" to have weak uniqueness?.. and use "type" as part of the key to enhance its uniqueness?

snazy replied:
We can't really enforce it, yes. It's what we get from the "outside system" i.e. Iceberg.
There will be some checks in higher level code that prevents reusing snapshot-IDs or schema-IDs though.

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java:
```diff
@@ -323,4 +323,35 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
+   * Retrieve the binary objects identified via {@code keys}. Binary objects that do not exist are
+   * not returned.
+   */
+  Stream<BinaryContent> getBinaries(Stream<BinaryKey> keys);
```
Well, `globalContent` effectively returns opaque binary data too... Could we use a different name for these objects? "Binaries" seems to be too generic to me :thinking:

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java:
```diff
@@ -323,4 +323,35 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
+   * <p>Either a "put-if-absent", if {@code expectedHash} is empty or a compare-and-swap based on
+   * the value of {@code expectedHash}. * *
+   *
+   * <p><b>IMPORTANT</b> keys of the {@link BinaryContent} binary objects must not be used via this
+   * method and {@link #putBinaries(Stream)}.
```
I'm not sure I understand this message, sorry :)

snazy replied:
Updated it.

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -96,18 +103,27 @@ public void eraseRepo() {
 
-  private List<Document> toIds(Collection<Hash> ids) {
-    return ids.stream().map(this::toId).collect(Collectors.toList());
+  private List<Document> toIdsStream(Stream<String> ids) {
```
nit: do we have to rename this method? It looks like there are no overloads :thinking:

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -482,4 +518,84 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
+            doc ->
+                client
+                    .getBlobs()
+                    .updateOne(
```
Should we check the result of `updateOne` (e.g. `result.wasAcknowledged()`)? I'm not sure Mongo Client will throw exceptions in all cases the update did not happen :thinking:

snazy replied:
done.

dimas-b reviewed (commented):
I have a very brief look at this PR. Sorry about naming nitpicks... will try to have a deeper review tomorrow.

snazy commented:
> Please correct me if I'm wrong: the overall idea is to have content objects versioned per branch with commit log and each of the versions can refer to zero or more "attachments" by their unique ID (`BinaryKey` or the renamed class). However, "attachments" themselves are immutable and not versioned... right?

Right. Those blobs are attached to a content on a commit. Multiple contents (for the same content id) can refer to the same attachment. Some are immutable (written via `DatabaseAdapter.putAttachments()`), some can be updated (via `DatabaseAdapter.consistentPutAttachment()`). The ones that can be updated represent some global state per content id - it just doesn't need the overall consistency guarantees that the current global-state w/ global-log + global-pointer provides.

tomekl007 commented on versioned/persist/dynamodb/src/main/java/org/projectnessie/versioned/persist/dynamodb/DynamoDatabaseAdapter.java:
```diff
@@ -468,4 +503,101 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
+      AttachmentKey key, AttachmentValue value, Optional<String> expectedVersion) {
+
+    AttributeValue newValueBytes =
+        AttributeValue.builder().b(SdkBytes.fromByteArray(value.toByteArray())).build();
```
why the `AttributeValue.builder` has those not descriptive methods like `s()` or `b()`?
Can we change those to some meaningful names?

snazy replied:
Ehm - no? That's AWS SDK

tomekl007 commented on versioned/persist/inmem/src/main/java/org/projectnessie/versioned/persist/inmem/InmemoryDatabaseAdapter.java:
```diff
@@ -252,4 +257,69 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
+                    return value.toByteString();
+                  }
+                } catch (InvalidProtocolBufferException e) {
+                  throw new RuntimeException(e);
```
shouldn't we create a custom exception that extends `RuntimeException` instead of using it directly?

snazy replied:
Generally, yes.
But this is the inmemory implementation, that's only used in tests (hope nobody uses it in prod...)

tomekl007 replied:
maybe add it in the Javadoc of this class (that it should not be used in prod)?

snazy replied:
Different topic ;)
If, then it should be everywhere - especially application.properties, website, etc. Wanna take a stab on that?

tomekl007 replied:
Maybe we should create a ticket for that and decide if it is worth doing depending on our priorities?

snazy replied:
Yup, please go ahead.

tomekl007 replied:
created: https://github.com/projectnessie/nessie/issues/3414

tomekl007 commented on versioned/persist/inmem/src/main/java/org/projectnessie/versioned/persist/inmem/InmemoryDatabaseAdapter.java:
```diff
@@ -252,4 +257,69 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
+      store.attachments.compute(
+          dbKey(key.toByteString()),
+          (k, current) -> {
+            if (expectedVersion.isPresent()) {
```
Do we have guidelines around handling the `Optional` type?
can we use `expectedVersion.map().orElse()` instead of `if(isPresent())...`?

snazy replied:
We don't have guidelines - and I think we shouldn't have.
IMO it should be "readable". As long as the "conventional" approach is readable (or "equally readable" compared to stream-ish one, it's fine, because the implicit one's less work for the VM.

tomekl007 replied:
I am asking because many OS projects have guidelines around handling Optional. A lot of them prefer the non-functional approach (`isPresent`) and some of them functional (`map()..`) so it maybe beneficial to have unified rules regarding `Optional` handling

tomekl007 commented on versioned/persist/inmem/src/main/java/org/projectnessie/versioned/persist/inmem/InmemoryDatabaseAdapter.java:
```diff
@@ -252,4 +257,69 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
+                }
+              }
+              // consistent-update failed, key not present or hash!=expected, throw
+              throw new IllegalStateException();
```
please add some message to this exception. I think it can be even:
`throw new IllegalStateException(consistent-update failed, key not present or hash!=expected)`

snazy replied:
It's a bit of a hack here. Adding a different approach to make it a bit clearer.

tomekl007 commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -229,26 +249,32 @@ private void delete(MongoCollection<Document> collection, Set<Hash> ids) {
-      MongoCollection<Document> collection, List<Hash> hashes, Function<Document, T> mapper) {
-    List<Document> ids = hashes.stream().map(this::toId).collect(Collectors.toList());
+  private <T> List<T> fetchMappedPageGeneric(
+      MongoCollection<Document> collection, List<String> idList, Function<Document, T> mapper) {
```
why do we get rid of type safety here (using `Hash`) and starting using `String` instead?
I see that the caller has a hash and transforms it to `String` (`Hash::asString`). 
Shouldn't we favor strong type instead of just passing `String`?

tomekl007 commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -485,4 +521,91 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
+            "Unacknowledged write to " + client.getAttachments().getNamespace());
+      }
+      return r.getMatchedCount() == 1L && r.getModifiedCount() == 1L;
+    } else {
```
do we want to execute the same else block for `getMatchedCount()>1` and `getMatchedCount()==0`?
Or should those two cases be handled separately?

snazy replied:
Not sure what you mean here wrt to `return` and the following `else`

tomekl007 commented on versioned/persist/serialize/src/main/java/org/projectnessie/versioned/persist/serialize/ProtoSerialization.java:
```diff
@@ -287,6 +293,126 @@ public static Key protoToKey(AdapterTypes.Key key) {
+        "%s::%s::%s", key.getContentId().getId(), key.getObjectType(), key.getObjectId());
+  }
+
+  public static AttachmentKey attachmentKeyFromString(String key) {
```
Should the methods that provide a construction of the `AttachmentKey` be in the `AttachmentKey.java` file?

snazy replied:
Tricky - `AttachmentKey` is a generated protobuf class

tomekl007 commented on versioned/persist/serialize/src/test/java/org/projectnessie/versioned/persist/serialize/TestSerialization.java:
```diff
@@ -360,6 +388,41 @@ public void stableOrderOfRepoDescProps() {
 
+  @Test
+  public void attachmentContentSerialization() throws Exception {
+    for (int i = 0; i < 500; i++) {
```
can you provide an explanation why `500` was chosen?

snazy replied:
I don't have a particular one - just to have some more randomness here.
500 is a bit too much - updated to 50

tomekl007 replied:
If we need randomness, maybe we should consider property-based testing?

tomekl007 commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -1684,4 +1709,129 @@ private void writeRefLog(Connection connection, AdapterTypes.RefLogEntry entry)
+    try {
+      String sql = sqlForManyPlaceholders(SqlStatements.SELECT_ATTACHMENTS, keysList.size());
+      try (PreparedStatement ps = conn.prepareStatement(sql)) {
+        ps.setString(1, config.getRepositoryId());
```
why not start from `0`?

snazy replied:
JDBC starts at 1

tomekl007 commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -1684,4 +1709,129 @@ private void writeRefLog(Connection connection, AdapterTypes.RefLogEntry entry)
+          ps.setString(3, config.getRepositoryId());
+          ps.setString(4, attachmentKeyAsString(ContentAttachmentKey.fromContent(attachment)));
+          ps.setString(5, expectedVersion.get());
+          if (ps.executeUpdate() != 1) {
```
I think that this block:
```
          if (ps.executeUpdate() != 1) {
            return false;
          }
```
can be moved before `ps.set..` and provide fast-path returning false without the need for setting the values in prepared statements, right?

snazy replied:
You're suggesting to execute the prepared-statements w/o all bind-variables configured?

tomekl007 replied:
ah, you are right. I thought that this value was pre-computed

tomekl007 commented on versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java:
```diff
@@ -1684,4 +1709,129 @@ private void writeRefLog(Connection connection, AdapterTypes.RefLogEntry entry)
+      txCommit(conn);
+    } catch (SQLException e) {
+      if (isIntegrityConstraintViolation(e)) {
+        return false;
```
shouldn't we log that? Right now the information about the failure will be hidden (because we are just returning `false`)

snazy commented on versioned/persist/inmem/src/main/java/org/projectnessie/versioned/persist/inmem/InmemoryDatabaseAdapter.java:
```diff
@@ -252,4 +257,69 @@ protected RefLog fetchFromRefLog(NonTransactionalOperationContext ctx, Hash refL
   }
+
+  @Override
+  protected void writeAttachments(Stream<Entry<AttachmentKey, AttachmentValue>> attachments) {
```
Those are in `org.projectnessie.versioned.persist.tests.AbstractBinary`

tomekl007 commented on versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractBinary.java:
```diff
@@ -0,0 +1,199 @@
+ * verify that the embedded and nested key-lists (think: full-key-lists in a commit-log-entry) work
+ * correctly.
+ */
+public abstract class AbstractBinary {
```
How do we separate units from integration tests?

snazy replied:
For the database adapters it's basically a "unit test", if the database adapter is in memory (so H2 + the inmem adapter) - if it's a separate process (e.g. Mongo, Dynamo, Postgres), then it's an integration test. Similar for the tests via Quarkus.

snazy replied:
IT = maven-failsafe-plugin
Test* or *Test = maven-surefire-plugin

You mean this?

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java:
```diff
@@ -302,6 +304,44 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
+   * the value of {@code expectedHash}. * *
+   *
+   * <p>Note: this method uses conditional put operations, unlike {@link #putAttachments(Stream)}.
+   * Should not use this method and {@link #putAttachments(Stream)} for the same keys.
```
Is this a hard requirement or more of a suggestion?

snazy replied:
The "layout" or "meaning" of the keys depends on the caller. The implementation would not complain, if the same key is used for both `consistentPutAttachment` and `putAttachments`, but the guarantees of `consistentPutAttachment` would be off.

dimas-b replied:
Makes sense. That was my understanding.. just wanted to confirm :)

dimas-b commented on versioned/persist/tx/src/test/java/org/projectnessie/versioned/persist/tx/AbstractTxDatabaseAdapterTest.java:
```diff
@@ -0,0 +1,61 @@
+
+public abstract class AbstractTxDatabaseAdapterTest extends AbstractDatabaseAdapterTest {
+  @Test
+  void isIntegrityConstraintViolation() throws Exception {
```
nit: rename to `insertOnConflict` perhaps (for clarity)?

snazy replied:
:+1:

snazy commented on versioned/persist/inmem/src/main/java/org/projectnessie/versioned/persist/inmem/InmemoryDatabaseAdapter.java:
```diff
@@ -262,4 +268,149 @@ protected List<RefLog> doFetchPageFromRefLog(
+        });
+  }
+
+  private void addAttachmentKey(AttachmentKey attachmentKey) {
```
done

snazy commented on versioned/persist/rocks/src/main/java/org/projectnessie/versioned/persist/rocks/RocksDatabaseAdapter.java:
```diff
@@ -455,4 +463,177 @@ private void checkForHashCollision(ColumnFamilyHandle cf, byte[] key)
+      if (keyList.isEmpty()) {
+        return Stream.empty();
+      }
+      List<ColumnFamilyHandle> handles =
+          IntStream.range(0, keyList.size())
+              .mapToObj(x -> dbInstance.getCfAttachments())
+              .collect(Collectors.toList());
```
:+1:

snazy commented on versioned/persist/serialize/src/main/java/org/projectnessie/versioned/persist/adapter/serialize/ProtoSerialization.java:
```diff
@@ -286,6 +293,134 @@ public static Key protoToKey(AdapterTypes.Key key) {
+    Preconditions.checkState(
+        !contentId.contains("::") && !objectType.contains("::") && !objectId.contains("::"),
+        "Attributes of an attachment key must not contain '::'");
+    return String.format("%s::%s::%s", contentId, objectType, objectId);
```
:+1:

snazy commented on versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractBinary.java:
```diff
@@ -0,0 +1,232 @@
+/**
+ * Verifies that a big-ish number of keys, split across multiple commits works and the correct
+ * results are returned for the commit-log, keys, global-states. This test is especially useful to
+ * verify that the embedded and nested key-lists (think: full-key-lists in a commit-log-entry) work
```
oops

snazy commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -492,6 +520,156 @@ protected List<RefLog> doFetchPageFromRefLog(
+
+    attachmentsKeys.forEach(
+        (cid, attachmentKeys) ->
+            verifyAcknowledged(
```
Yea - I thought about this "nosql problem" quite a bit.
I suspect, there's no "always correct" solution here. OTOH it shouldn't matter, beside there might be rows in the content-attachments table that are never referenced.
If a write fails, the whole commit operation fails and the attachments, if not the same in another commit, will be just garbage in the database.
Trying each key individually might be too slow.

dimas-b replied:
re: failures - my impression was that the attachment key list is by content ID, not by commit, so I though losing keys might have wider implications than just one commit. WDYT?

snazy replied:
Duh - we don't do batch-writes anywhere for Mongo. Shall we tackle that as a follow-up and fix it everywhere?

dimas-b replied:
sure :+1:

snazy replied:
> re: failures - my impression was that the attachment key list is by content ID, not by commit, so I thought losing keys might have wider implications than just one commit. WDYT?

Yes, attachments are per-content-ID. But it's the other way around: the "list of attachment keys" won't have missing entries, but potentially "too many". But never have a missing attachment (unless it's been explicitly deleted).

snazy replied:
^ that's to "work around" the issue w/ no-sql databases / no (cheap-ish) "multi-key" transactions (aka "all or nothing")

snazy replied:
Follow-up for the batch-writes here: #4159

snazy commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -100,18 +109,28 @@ public void eraseRepo() {
     // Note: the order of `put` calls matters
     idDoc.put(ID_REPO_NAME, repositoryId);
-    idDoc.put(ID_HASH_NAME, id.asString());
+    idDoc.put(ID_HASH_NAME, id);
```
Wanted to keep the behavior the same for all database adapters. It's tricky to handle decomposed keys in some - especially in SQL :(

dimas-b replied:
Oh, I meant using decomposed keys only in Mongo, where it would be more meaningful, I hope.

dimas-b replied:
Adapters receive the full key as a java object... would it be too much inconvenience to have different attachment key representation in some adapters (i.e. Mongo)?

snazy replied:
What do you have in mind for Mongo-only? Only thing that makes sense to split out would be the content-ID. But on the read-path, wouldn't it be better to "distribute" attachment-reads - i.e. not consider the content-ID as part of the "sharding-key" (or however it's called)?

dimas-b replied:
If we use opaque IDs (current state) then the whole key body will be used for sharding. If we use a composite key we can defer sharding decisions to deployment time, I think.

snazy replied:
added that as a separate commit

snazy commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java:
```diff
@@ -302,6 +304,44 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
+   * Retrieve the content attachments identified via {@code keys}. Attachments that do not exist are
+   * not returned.
+   */
+  Stream<ContentAttachment> getAttachments(Stream<ContentAttachmentKey> keys);
```
:+1: 
Added a note that it's up to the implementation whether the implemetnation "terminates" the input stream or not.

snazy commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -492,6 +520,156 @@ protected List<RefLog> doFetchPageFromRefLog(
+      Stream<AttachmentKey> keys) {
+    return fetchMappedPageGeneric(
+            client.getAttachments(),
+            keys.map(ProtoSerialization::attachmentKeyAsString).collect(Collectors.toList()),
```
Hm - Mongo doesn't seem to complain with even 10000 keys here. But... Dynamo does: `'requestItems.attachments.member.keys' failed to satisfy constraint: Member must have length less than or equal to 100`
Added generic paging.

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -224,26 +246,32 @@ private <T, ID> T loadById(MongoCollection<Document> collection, ID id, Parser<T
+        collection.find(Filters.in(ID_PROPERTY_NAME, ids)).limit(idList.size());
 
-    HashMap<Hash, Document> loaded = new HashMap<>(hashes.size() * 4 / 3 + 1, 0.75f);
+    HashMap<String, Document> loaded = new HashMap<>(idList.size() * 4 / 3 + 1, 0.75f);
```
nit: `com.google.common.collect.Maps.newHashMapWithExpectedSize(N)`?

dimas-b commented on site/docs/try/configuration.md:
```diff
@@ -86,6 +86,7 @@ The following configurations are advanced configurations to configure how Nessie
 | `nessie.version.store.advanced.commit-timeout`                  | `500`               | `int`    | Sets the timeout for CAS-like operations in milliseconds.                                                                                                                                                                             |
 | `nessie.version.store.advanced.commit-retries`                  | `Integer.MAX_VALUE` | `int`    | Sets the maximum retries for CAS-like operations.                                                                                                                                                                                     |
 | `nessie.version.store.advanced.nontx.parents-per-global-commit` | `50`                | `int`    | Sets the number of parent-global-commit-hashes.                                                                                                                                                                                       |
+| `nessie.version.store.advanced.attachment-keys-batch-size`      | `100`               | `int`    | Sets the number of content attachments that are written or retrieved at once. Some implementations may silently adopt this value to database limits or implementation requirements.                                                   |
```
`adopt` -> `adapt`?

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -492,6 +520,156 @@ protected List<RefLog> doFetchPageFromRefLog(
+      return Stream.empty();
+    }
+    List<String> keys = doc.getList(DATA_PROPERTY_NAME, String.class);
+    return keys.stream().distinct().map(ProtoSerialization::attachmentKeyFromString);
```
Do we really need `.distinct()` since we use Mongo set operations for writing this data?

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -492,6 +520,156 @@ protected List<RefLog> doFetchPageFromRefLog(
+    if (doc == null) {
+      return Stream.empty();
+    }
+    List<String> keys = doc.getList(DATA_PROPERTY_NAME, String.class);
```
Would it be worth trying to fetch the inner property (Mongo projection) and rely on Mongo's `FindIterable` to do the streaming?

snazy replied:
I doubt

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -68,9 +76,13 @@
   private static final String ID_HASH_NAME = "hash";
+  private static final String ID_ATT_CONTENT_ID = "cid";
+  private static final String ID_ATT_OBJECT_TYPE = "type";
+  private static final String ID_ATT_OBJECT_ID = "oid";
```
nit: I believe "attribute" is usually abbreviated as "attr" :)

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -100,20 +112,52 @@ public void eraseRepo() {
+    return toId(id.asString());
+  }
+
+  private Document toId(String id) {
```
Do we need this change anymore? It looks like `toId(String)` is called only from `toId(Hash)` :thinking:

dimas-b commented on versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:
```diff
@@ -100,20 +112,52 @@ public void eraseRepo() {
+    return idDoc;
+  }
+
+  private List<Document> toIdsFromStrings(Collection<Hash> ids) {
```
hmmm, the name has "from Strings" but the arg is a collection of `Hash`s?

nastra commented:
will review the PR tomorrow, so please don't merge ;)

snazy commented:
Changes since the last review round:
* Attachment values persisted with "format" _and_ "compression" attributes
* Attachment values field "object ID", which represents e.g. the snapshot ID or schema ID to handle the case when the same content ID has different e.g. schemas or partition-specs or sort-orders on different branches but with the same schema-ID/partition-spec-ID/sort-order-ID

Attachment data serialization format can be JSON or CBOR - the latter wastes less space.

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java:
```diff
@@ -291,6 +293,56 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
+  /**
+   * Consistent put-attachment operation.
+   *
+   * <p>Either a "put-if-absent", if {@code expectedHash} is empty or a compare-and-swap based on
```
`expectedHash` -> `expectedVersion`?

dimas-b commented on versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/DatabaseAdapter.java:
```diff
@@ -291,6 +293,56 @@ void updateRepositoryDescription(Function<RepoDescription, RepoDescription> upda
+   * @param attachment the attachment to write
+   * @param expectedVersion indicator for put-if-absent or the expected value on an existing item
+   */
+  boolean consistentPutAttachment(ContentAttachment attachment, Optional<String> expectedVersion);
```
Just wondering if we may need change log for attachments? Since the key is probably going to be frozen into a commit, could it be necessary for us to get previous versions of attachments when/after changes are made? Since we track top-level content changes in commits, should we also track attachment changes for the same of having full history?

snazy replied:
Not sure TBH. We do not need this consisten-put at the moment. It was intended to potentially allow sequence-numbers per content-id - like column-IDs and such, if we want to at some point.

dimas-b reviewed (approved):
Sorry, some last minute comments :) ... but minor / deferable.
